lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -554,8 +555,13 @@ applyOurTxToUTxO
     -> UTxO
     -> Maybe ((Tx, TxMeta), UTxO)
 applyOurTxToUTxO !slotNo !blockHeight !s !tx !prevUTxO =
-    if hasKnownWithdrawal || prevUTxO /= ourNextUTxO
-        then Just ((tx {fee = actualFee dir}, txmeta), ourNextUTxO)
+    if ourWithdrawalSum /= mempty || prevUTxO /= ourNextUTxO
+        then
+            let updatedTx = tx
+                    { fee = actualFee dir
+                    , outputs = filter (ours s . address) (outputs tx)
```

Comment:
Yes this looks like it will work for `listAssets`.
If everything breaks due to this change, then perhaps we could annotate the TxOut address about whether it belongs to the wallet.

Reply:
Uhh, this does indeed look like it could break everything. 😅 I think the original idea for the `Tx` type was to represent a transaction exactly as it appears on the chain — no modifications. (I believe that the adjustment to the `fee` field is a hack, as this field did not exist in Byron, but is added for clarity.) Changing this assumption has several implications:

* The `/v2/wallets/{walletId}/transactions` endpoint changes in a backwards-incompatible way and no longer presents transactions as they appear on-chain. For example, `cardano-node` will return a different result than this endpoint if you give it the same transaction hash.
* `Tx` are stored in the database via `putTxHistory`. If we change the contents of `Tx`, we would have to add a migration for the database as well.
* The `inputIx` field of the `TxIn` refers to an index in the list `outputs`. If we filter the list, the indices get rearranged. 😱 We are safe in that `ourNextUTxO` uses the original `tx`, not the `updatedTx` to compute the `UTxO` set. Users of the `/v2/…/transactions` endpoint are not so lucky. If we really want to filter outputs from the `Tx`, we should at least record their indices as well.

In light of these implications, my preference would be to keep the original idea of `Tx` representing a complete transaction, and do the filtering in `listAssets` a posteriori.

Reply:
The `Tx` type in cardano-wallet has never been a transaction exactly as it appears on the chain. Multiple fields are missing!
It can be whatever we want or need it to be. We can compute info about the transaction once during restoration, then cache it in the db. An example of such info is the transaction amount.

I ask, do users really need to see TxOuts which don't concern their wallet?
If they do, perhaps the wallet should store the original serialized transaction in the database for this purpose.

Although, if our code is making assumptions all over the place that the TxOuts aren't filtered -- you raise a very good point about numbering -- then it's probably easiest to _not_ filter, but instead annotate the TxOut address with a flag indicating whether the address belongs to this wallet.

Yes this will probably need database migrations; we can do migrations.
Also users will not see a propertly filtered listAssets unless they restore from genesis. This is also fine.

Reply:
@HeinrichApfelmus appreciate your input! 
I've added *Solution #2* to the issue description that takes your input into account as described by @rvl